BrowserKit/Sources/TabDataStore/TabDataStore.swift

```diff
@@ -96,20 +125,38 @@ public actor DefaultTabDataStore: TabDataStore {
     }
 
     // MARK: Saving Data
-    public func saveWindowData(window: WindowData) async {
+    public func saveWindowDataWithBackup(window: WindowData) async {
         Task {
-            if let windowSavingPath = self.windowURLPath(for: window.id) {
+            if let windowSavingPath = self.windowURLPath(for: window.id),
+               let backupWindowSavingPath = self.windowBackupURLPath(for: window.id) {
                 do {
+                    if self.checkIfFileExistsAtPath(path: windowSavingPath) {
+                        do {
+                            try DefaultTabDataStore.tabFileManager.copyItem(at: windowSavingPath, to: backupWindowSavingPath)
+                        } catch {
+                            self.logger.log("Failed to create window data backup: \(error)", level: .debug, category: .tabs)
+                        }
+                    }
                     try await self.writeWindowData(windowData: window, to: windowSavingPath)
                 } catch {
-                    self.logger.log("Failed to save window data: \(error)",
-                                    level: .debug,
-                                    category: .tabs)
+                    self.logger.log("Failed to save window data: \(error)", level: .debug, category: .tabs)
+                    if self.checkIfFileExistsAtPath(path: backupWindowSavingPath) {
+                        do {
+                            try DefaultTabDataStore.tabFileManager.removeItem(at: windowSavingPath)
+                            try DefaultTabDataStore.tabFileManager.moveItem(at: backupWindowSavingPath, to: windowSavingPath)
+                        } catch {
+                            self.logger.log("Failed to restore backup: \(error)", level: .debug, category: .tabs)
+                        }
+                    }
```

Comment:
I don't think this block of code is needed here. In a normal use case we will attempt to write again soon and if that also fails then we can determine if the backup is needed when restoring at start up.